protocols/kad/src/query/peers/closest.rs

```diff
@@ -691,4 +693,77 @@ mod tests {
 
         QuickCheck::new().tests(10).quickcheck(prop as fn(_) -> _)
     }
+
+    #[test]
+    fn stalled_iter_at_capacity_use_max_of_parallelism_and_num_results() {
+        let target: KeyBytes = Key::from(PeerId::random()).into();
+
+        let mut iter = ClosestPeersIter::with_config(
+            ClosestPeersIterConfig {
+                num_results: 4,
+                parallelism: 8,
+                ..ClosestPeersIterConfig::default()
+            },
+            target,
+            std::iter::empty(),
+        );
+        iter.state = State::Stalled;
+        iter.num_waiting = 7; // Smaller than `parallelism`, but larger than `num_results`.
+
+        assert!(!iter.at_capacity());
+    }
+
+    /// When not making progress for `parallelism` time a [`ClosestPeersIter`] becomes
```

Comment:
This test is likely a bit over the top. I left it here to showcase the issue in a more in-depth scenario. I would suggest removing it before merging. What do you think?

Reply:
Yes, I don't think it adds much to the other test, especially if the other is turned into a property test. Thanks for writing it though.